support/src/main/java/org/jenkinsci/plugins/workflow/support/steps/MilestoneStepExecution.java

```diff
@@ -0,0 +1,356 @@
+package org.jenkinsci.plugins.workflow.support.steps;
+
+import static java.util.logging.Level.WARNING;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.Map.Entry;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+
+import org.jenkinsci.plugins.workflow.flow.FlowExecutionOwner;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.steps.AbstractStepExecutionImpl;
+import org.jenkinsci.plugins.workflow.steps.FlowInterruptedException;
+import org.jenkinsci.plugins.workflow.steps.StepContext;
+import org.jenkinsci.plugins.workflow.steps.StepContextParameter;
+
+import com.google.inject.Inject;
+
+import hudson.Extension;
+import hudson.XmlFile;
+import hudson.model.Job;
+import hudson.model.Result;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.model.listeners.RunListener;
+import jenkins.model.CauseOfInterruption;
+import jenkins.model.Jenkins;
+
+public class MilestoneStepExecution extends AbstractStepExecutionImpl {
+
+    private static final Logger LOGGER = Logger.getLogger(MilestoneStepExecution.class.getName());
+
+    @Inject(optional=true) private transient MilestoneStep step;
+    @StepContextParameter private transient Run<?,?> run;
+    @StepContextParameter private transient FlowNode node;
+    @StepContextParameter private transient TaskListener listener;
+
+    private static Map<String, Map<Integer, Milestone>> milestonesByOrdinalByJob;
```

Comment:
Static context has a tendency to behave iffy in Jenkins. Have you considered putting this as a member of the Descriptor who's life cycle is controlled by Jenkins?